cmd/state-svc/internal/resolver/analytics/analytics.go

```diff
@@ -219,12 +227,52 @@ func (r *Resolver) flush() error {
 		return errs.Wrap(err, "Failed to load deferred events")
 	}
 	for _, event := range events {
-		r.event(event)
+		ev := &event
+		r.addProjectIDToEvent(ev)
+		r.event(ev)
 	}
 
 	return nil
 }
 
+func (r *Resolver) addProjectIDToEvent(ev *event.EventData) {
+	if ev.ProjectName == "" || ev.ProjectID != "" {
+		return
+	}
+
+	ev.ProjectID = r.projectID(ev.ProjectName)
+}
+
+// projectID resolves the projectID from projectName and caches the result in the provided projectIDMap
+func (r *Resolver) projectID(projectName string) string {
+	if projectName == "" {
+		return ""
+	}
+
+	// Lock mutex to prevent resolving the same projectName more than once
+	r.projectIDMutex.Lock()
+	defer r.projectIDMutex.Unlock()
+
+	if pi, ok := r.projectIDCache.Get(projectName); ok {
+		return pi.(string)
+	}
+
+	pn, err := project.ParseNamespace(projectName)
+	if err != nil {
+		logging.Error("Failed to parse project namespace %s: %s", projectName, errs.JoinMessage(err))
+	}
+
+	pj, err := model.FetchProjectByName(pn.Owner, pn.Project)
+	if err != nil {
+		logging.Error("Failed get project by name: %s", errs.JoinMessage(err))
+	}
+
+	pi := string(pj.ProjectID)
+	r.projectIDCache.Set(projectName, pi, cache.DefaultExpiration)
```

Comment:
`cache.DefaultExpiration` appears to be `0`:

https://github.com/ActiveState/cli/blob/55c0ada4317b700374fff8c780d56b657c226f03/vendor/github.com/patrickmn/go-cache/cache.go#L32

Either way I'd prefer we don't rely on a default that's defined by a third party for something like this.

Reply:
The `DefaultExpiration` is the the first argument to the `cache.New()` function.  That way you have to set it in only one place.  It is not determined by the package.